### PR TITLE
Add compatibility testing for JDBC driver

### LINKED ISSUE #56722: Support compatibility test for JDBC driver
Forwards compatibility was added to our JDBC driver starting in version 7.7.0 (see https://github.com/elastic/elasticsearch/pull/56223). The compatibility guarantee being that for any given version of the driver, it should support (mostly) communication for future versions of Elasticsearch of the same major version. For example, version 7.7.0 of the driver should be able to talk to Elasticearch nodes of version 7.7.1, 7.8.0, 7.8.1, 7.9.0 and so on...

Currently, there is no automated testing in place to ensure we maintain forwards compatibility. It's also not completely clear where, or how these tests should run. Currently, the build only supports _backwards_ compatibility. That is, testing branch code against assets or artifacts from earlier versions or branches. Building or resolving artifacts to test against from _future_ branches currently isn't supported, since, for example, the `7.x` branch has no knowledge of the code in `master`.

We'll need to sort out a few things here, such as which _direction_ we test in. Do we do backwards testing, where in which future branches grab the older JDBC drivers and run the tests, or forwards testing, where older branches grab the new Elasticsearch distribution and run the tests? Much of this is implementation detail, but it has some semantic impact, as with the former method, a failure in a given branch generally points to an incompatible change made in a different branch (but not necessarily). Going the latter route would require more effort, since we'd need to add support to build unreleased artifacts from future branches, which we currently don't do.

cc @matriv

### PR DESCRIPTION
**Pull request by mark-vieira**
This commit adds compatibility testing of our JDBC driver against
different Elasticsearch versions. Although we are really testing the
forwards compatibility nature of the JDBC driver we model the testing
the same as we do existing BWC tests, that is, with the current branch
fetching the earlier versions of the artifact that is to be tested. In
this case, that's the JDBC driver itself.

Because the tests include the JDBC driver jar on it's classpath we had
to change the packaging of the driver jar in order to avoid jarhell and
other conflicting dependency issues when using an old JDBC driver with
later branches. For this we simply relocate all driver dependencies in
the shadow jar under a "shadowed" package. This allows the JDBC driver
to use the correct version of Elasticsearch libs classes, while the
tests themselves use their versions. Since this required a change to the
driver jar compatibility testing can only go back as far as that version
which at the time of this commit is 7.8.1.

We are starting with merging this into the 7.8 branch since due to the
"forwards" nature of this change we have to do it in reverse order we'd
normally merge PRs in. So with the changes to the JDBC jar packaging
in 7.8, we'll be able to "forwardport" this to 7.x and then finally to master.

Closes #56722

### CODE REVIEW

**Comment by mark-vieira** [comment]
Note, that because this branch represents 7.8.1 and this is also the first versions to support compatibility testing, there won't actually be any BWC tests executed for JDBC here. Once this is merged back into 7.x we'll start seeing compatibility tests run.

**Review by matriv — APPROVED** [review]
LGTM. Thank you @mark-vieira for this effort!

File: x-pack/plugin/sql/qa/jdbc/build.gradle
Comment (mark-vieira) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: The intention was that this class contains the logic for determining types of compatibility. So index, wire, etc. The thought is that JDBC driver compatibility is just another form of that, so why deviate from the pattern and bury that in the JDBC QA project? The class is still generic by that definition since doesn't actually reference any Gradle projects.

Comment (rjernst) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: Except in this case since we have just recently started compatibility support, when would it deviate from index compatible versions (also, is index compatible correct?)? When we branch 8.x, wouldn't this in master be compatible with all index compatible versions? I don't see this as another type of compatibility, I see it as another feature of the system that should be aligned with existing compatibility rules.

Comment (mark-vieira) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: I can't comment on whether or not JDBC driver compatibility should strictly align with index compatibility or not, you'd have to reach out to the SQL team for that. But in my mind, even if it is _essentially_ the same, it didn't feel right to highjack that definition. The introduction of JDBC compatibility _is_ new, and the semantics might change, so I thought it best to be explicit. @matriv perhaps you could clarify exactly what the JDBC driver compatibility constraints will be going forward?

Also, FWIW, there are other definitions of compatibility that deviate from index/wire compatibility. We are having this discussion right now regarding things like cross cluster search. I suspect once we have a concrete definition and implement proper testing that's we'd similarly define the set of compatible versions in `BwcVersions` as that's what I interpret it's purpose is for. That is, to answer the question "what versions of Elasticsearch should scenario/feature X work with?".

Comment (mark-vieira) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: To add to that, REST compatibility testing falls in the same bucket. The versions tested will likely be identical to index compatibility, but it seems more correct to again model it as a separate thing, at least until we one day arrive at some kind of universal "client compatible" definition, which seems a ways off.

Comment (matriv) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: JDBC is supposed to be fwds compatible past the major version (8.0.0), so we shouldn't confuse/connect this compatibility with the index compatibility.

Comment (mark-vieira) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: So to confirm, when 9.0 is branched, it should continue to work with 7.x drivers? How far will this go? This is going to add quite a bit of testing, but I guess we can cross that bridge when we come to it and potentially shred JDBC compat testing into it's own job.

With that in mind though you are right, we shouldn't base this on index compatibility. I'll fix this to simply return all versions starting with 7.8.1.

Comment (matriv) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: The idea in general is yes, to always provide that fwd compatibility as most jdbc drivers do.
They can add new features/optimisations but they shouldn't break existing functionality when talking to a newer server (ES node/cluster).

Comment (mark-vieira) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: This all sounds a bit fuzzy. Certainly we don't intent to maintain compatibility in future versions of elasticsearch inevitably. I mean, we _do_ change even REST APIs at major version boundaries to some extent. @matriv where are the compatibility guarantees for the driver documented? What exactly are we telling users?

Comment (bpintea) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: The JDBC driver (and any other SQL client) will want to align with the rest of the ES clients in regards to forward compatibility. I believe the current policy is for the server to remain `$major - 1` backwards compatible with the clients (which is also in line with #51816 and the SQL targets). 
So a 7.x JDBC wouldn't need to be compatible with any 9.x ES, just 8.x.

Comment (matriv) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: @bpintea Thx for clarifying, I didn't remember the `major - 1`.

Comment (mark-vieira) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: Can someone point me to documentation on this? I don't think this is a global compatibility policy with regards to clients, especially since the REST compatibility layer is not yet in place which would allow a 7.x client to talk to an 8.x server. I'm not even sure there is much in the way of guarantees for minor versions.

Sorry if I'm being pedantic here but we run into this compatibility discussion all the time and folks generally give fuzzy answers. If we are adding testing to ensure a certain level of compatibility, then that should be well documented and communicated to users (and developers). I'm not sure how we can expect users of the client, or engineers working on elasticsearch to be on the same page if what @bpintea describes doesn't seem to be written down anywhere.

@matriv is there a plan to document client forward compatibility? Has this been done but I just can't find it anywhere?

Comment (bpintea) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: > Can someone point me to documentation on this? I don't think this is a global compatibility policy with regards to clients, especially since the REST compatibility layer is not yet in place which would allow a 7.x client to talk to an 8.x server.

Unfortunately I'm not sure of any documentation on this either and neither sure if this is coming with guarantees or just a best effort. Maybe @colings86 could help us with it?

> I'm not sure how we can expect users of the client, or engineers working on elasticsearch to be on the same page if what @bpintea describes doesn't seem to be written down anywhere.

As far as the other ES clients are concerned, I believe that the implementation of #51816 will likely make this clear and documented. 
SQL backwards compatibility effort started along #51816, but we need to have it in place sooner than 8.x due to 3rd party integration requirements. All the SQL engineering team is of course aware of it and as far as the users are concerned, up to 7.6 we've required strict server-client lock-step versioning. So there should hopefully be no disruption and no false assumptions on either side.

> is there a plan to document client forward compatibility?

There's no plan yet since it depends on the testing too. We've been holding off pushing obviously breaking changes within the 7.x branch past 7.6 (at SQL level, not REST API) since we don't have the testing in place. I believe once we have it, we'll be able to publish this policy too.
Hope that this clarifies it a bit more and I think we can formalise it further if needed to get this work further.

Comment (mark-vieira) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: Thanks Bogdan. With that said I suppose the current implementation is fine since major-1 is effectively the same as index compatibility. I'm still advocating for this to live in BWCVersions for now since that class's job is to describe compatibility.

Comment (matriv) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: That sounds good to me.

Comment (rjernst) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: @mark-vieira I agree it is the job of BwcVersions to describe compatibility. But as you say, major-1 *is* the same as index compatibility. Ignoring the names, we effectively have 2 compatibility sets in ES, both described in code: index compatibility (major - 1), and wire compatibility (major-1.last). My push back here is to creating extra concepts that don't need to exist. I think the compatibility of any part of the system should match with one of those two concepts. Thus far we've called it index compatibility in this case. Giving jdbc it's own version set here makes it confusing both because it would not exist in code outside gradle (see Version.java in elasticsearch.jar) and diverging from one of those two concepts would add cognitive load to users and developers in understanding compatibility. Thus, I still think we should avoid creating a new set here which (after the initial version) should not differ from one of those two sets.

Comment (mark-vieira) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: I'm perfectly content with that if we agree that client compatibility = index compatibility and that we agree to maintain that going forward. So far we've been discussion client compatibility as it's own thing, and only incidentally does it overlap with index compatibility. That is, the decision to make ensure client compatibility across the versions we do isn't so much to align with index compatibility, but really to support upgrades in a similar way to index compatibility. By that definition "index compatibility" is really a poor name but we don't have to address that here.

@bpintea Does what I saw above make sense? That is, the use cases we wish to address with "major - 1" compatibility with clients is the same as the reason we ensure index compatibility across the same versions. That is, to make sure that folks can go from any minor version of a given major, to any minor version of the next major without having to do any intermediary upgrades (as is the case for zero-downtime rolling upgrades).

Comment (matriv) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: I'd say that since currently the scheme we support regarding JDBC fwds compatibility matches the one about index compatibility we can proceed with that and if something changes re-open this discussion in the future.

Comment (bpintea) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: > @bpintea Does what I saw above make sense? That is, the use cases we wish to address with "major - 1" compatibility with clients is the same as the reason we ensure index compatibility across the same versions.

Sure, it does.
And while not sure if they don't actually stem from the same place, I'm actually glad that the index bwc policy overlaps with clients', hoping for easier maintenance and better testing coverage.

Comment (mark-vieira) [reply x-pack/plugin/sql/qa/jdbc/build.gradle:74]: Cool. I'll push forward with that then. That said, I think we've missed the boat for getting this in 7.8. Sorry for the delay, I just think it makes sense for us all to be on the same page here. Although we've already branched 7.9 I'm going to get this merged in there, as the JDBC jar changes are necessary in order to have testing in 7.x. So I'll close this PR for now and open a new one against 7.9.

@rjernst Given the clarifications from @matriv and @bpintea I'm good with piggybacking on index compatibility as the definition for this as well. I'll modify the 7.9 PR appropriately for that.

**Pull request closed without merging**
